libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/publisher/CordaPublisherImpl.kt

```diff
@@ -282,15 +284,16 @@ internal class CordaPublisherImpl(
      */
     private fun resetProducer() {
         closeProducerAndSuppressExceptions()
-        cordaProducer = cordaProducerBuilder.createProducer(producerConfig, config.messageBusConfig)
+        cordaProducer.set(cordaProducerBuilder.createProducer(producerConfig, config.messageBusConfig))
     }
 
     /**
      * If an exception is thrown whilst closing, swallow the error to avoid double exceptions.
      */
     private fun closeProducerAndSuppressExceptions() {
         try {
-            cordaProducer.close()
+            cordaProducer.get()?.close()
```

Comment:
Same here, another thread scheduled in at this point (between lines 295 and 296) would be operating on a closed `cordaProducer` before this thread set is to null, so it appears to be half a solution. Perhaps there's some piece of information I'm missing?

Reply:
I think these two lines could be resolved by
```suggestion
            cordaProducer.getAndSet(null)?.close()
```

Reply:
Though, in hindsight, is there any use to making it nullable?

Reply:
Well it does offer limited protection against operating on a closed producer, except at line 193.

I think if we believe this class should be thread safe, we should resolve it everywhere. But I'm not sure why making this thread safe fixes the issue at the moment because I don't know enough about it.

Reply:
The problem, IIRC, is that we are closing and recreating a publisher, but some thread will, inevitably, access the publisher in the meantime.  And so Kafka ring-fences the underlying producer.  Then, when we get that error, it stops the re-creation too.  So I _think_ we end up stuck from that point.  Syncronising all access to the producer (especially around re-creation) should fix this, I think.

Reply:
I agree with @fowlerrr that having a lock on producer should fix the issue, but I'm worried about performance decrease. Currently we don't have an environment to test performance before and after changes.